### PR TITLE
[OAP-1824][SQL data source cache] Decouple DataSourceScanExec.scala

### PR DESCRIPTION
**Pull request by haojinIntel**
Create OapFileSourceScanExec to replace FileSourceScanExec.

### CODE REVIEW

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/OapFileSourceStrategy.scala [hunk 1]
```diff
@@ -33,10 +32,10 @@ object OapFileSourceStrategy extends Strategy with Logging {
 
     /**
      * After call [[FileSourceStrategy]], the result [[SparkPlan]] should the follow 4 scenarios:
-     *  1. [[ProjectExec]] -> [[FilterExec]] -> [[FileSourceScanExec]]
-     *  2. [[ProjectExec]] -> [[FileSourceScanExec]]
-     *  3. [[FilterExec]] -> [[FileSourceScanExec]]
-     *  4. [[FileSourceScanExec]]
+     *  1. [[ProjectExec]] -> [[FilterExec]] -> [[OapFileSourceScanExec]]
+     *  2. [[ProjectExec]] -> [[OapFileSourceScanExec]]
+     *  3. [[FilterExec]] -> [[OapFileSourceScanExec]]
```
Comment (zhixingheyi-tian) [inline oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/OapFileSourceStrategy.scala:37]: Should  leave this comment intact. no change

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/OapFileSourceScanExec.scala [hunk 2]
```diff
@@ -17,135 +17,28 @@
...
 
-import org.apache.commons.lang3.StringUtils
 import org.apache.hadoop.fs.Path
 
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.{InternalRow, TableIdentifier}
 import org.apache.spark.sql.catalyst.catalog.BucketSpec
-import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.{And, Ascending, Attribute, AttributeReference, BoundReference, Expression, PlanExpression, Predicate, SortOrder, UnsafeProjection}
```
Comment (LuciferYang) [inline oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/OapFileSourceScanExec.scala:30]: `import org.apache.spark.sql.catalyst.expressions._`

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/OapFileSourceStrategy.scala [hunk 3]
```diff
@@ -21,8 +21,7 @@ import org.apache.spark.internal.Logging
 import org.apache.spark.sql.{execution, SparkSession, Strategy}
 import org.apache.spark.sql.catalyst.planning.PhysicalOperation
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
-import org.apache.spark.sql.execution.{FileSourceScanExec, FilterExec, ProjectExec, SparkPlan}
-import org.apache.spark.sql.execution.dynamicpruning.PlanDynamicPruningFilters
+import org.apache.spark.sql.execution.{FileSourceScanExec, FilterExec, OapFileSourceScanExec, ProjectExec, SparkPlan}
```
Comment (LuciferYang) [inline oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/OapFileSourceStrategy.scala:24]: seems `FileSourceScanExec` is redundant

File: oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/CacheHotTablesSuite.scala [hunk 4]
```diff
@@ -72,10 +70,18 @@ abstract class CacheHotTablesSuite extends QueryTest with SharedOapContext with
...
-    val relation = scan.asInstanceOf[FileSourceScanExec].relation
-    assert(relation.isInstanceOf[HadoopFsRelation])
-    assert(verifyFileFormat(relation.fileFormat))
+    var relation: HadoopFsRelation = null
+    try {
+      assert(scan.isInstanceOf[OapFileSourceScanExec])
+      relation = scan.asInstanceOf[OapFileSourceScanExec].relation
+    } catch {
+      case ex: Exception =>
+        assert(scan.isInstanceOf[FileSourceScanExec])
```
Comment (LuciferYang) [inline oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/CacheHotTablesSuite.scala:79]: A little confused, why add this branch? I think we should reorganize test to ensure the `relation ` is `OapFileSourceScanExec`

File: oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapFileSourceStrategySuite.scala [hunk 5]
```diff
@@ -78,10 +78,18 @@ abstract class OapFileSourceStrategySuite extends QueryTest with SharedOapContex
...
-      assert(scan.isInstanceOf[FileSourceScanExec])
-      val relation = scan.asInstanceOf[FileSourceScanExec].relation
-      assert(relation.isInstanceOf[HadoopFsRelation])
-      assert(verifyFileFormat(relation.fileFormat))
+      var relation: HadoopFsRelation = null
+      try {
+        assert(scan.isInstanceOf[OapFileSourceScanExec])
+        relation = scan.asInstanceOf[OapFileSourceScanExec].relation
+      } catch {
+        case ex: Exception =>
```
Comment (LuciferYang) [inline oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapFileSourceStrategySuite.scala:86]: ditto

File: oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/CacheHotTablesSuite.scala [hunk 6]
```diff
@@ -101,10 +107,18 @@ abstract class CacheHotTablesSuite extends QueryTest with SharedOapContext with
...
-    assert(scan.isInstanceOf[FileSourceScanExec])
-    val relation = scan.asInstanceOf[FileSourceScanExec].relation
-    assert(relation.isInstanceOf[HadoopFsRelation])
-    assert(verifyFileFormat(relation.fileFormat))
+    var relation: HadoopFsRelation = null
+    try {
+      assert(scan.isInstanceOf[OapFileSourceScanExec])
+      relation = scan.asInstanceOf[OapFileSourceScanExec].relation
+    } catch {
+      case ex: Exception =>
```
Comment (LuciferYang) [inline oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/CacheHotTablesSuite.scala:115]: ditto

File: oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapFileSourceStrategySuite.scala [hunk 7]
```diff
@@ -107,10 +115,18 @@ abstract class OapFileSourceStrategySuite extends QueryTest with SharedOapContex
...
-    assert(scan.isInstanceOf[FileSourceScanExec])
-    val relation = scan.asInstanceOf[FileSourceScanExec].relation
-    assert(relation.isInstanceOf[HadoopFsRelation])
-    assert(verifyFileFormat(relation.fileFormat))
+    var relation: HadoopFsRelation = null
+    try {
+      assert(scan.isInstanceOf[OapFileSourceScanExec])
+      relation = scan.asInstanceOf[OapFileSourceScanExec].relation
+    } catch {
+      case ex: Exception =>
```
Comment (LuciferYang) [inline oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapFileSourceStrategySuite.scala:123]: ditto

**Review by LuciferYang — COMMENTED** [review]
Good for me and leave some comments.

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/OapFileSourceScanExec.scala (hunk 2 above)
Comment (haojinIntel) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/OapFileSourceScanExec.scala]: Refined.

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/OapFileSourceStrategy.scala (hunk 3 above)
Comment (haojinIntel) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/OapFileSourceStrategy.scala:24]: FileSourceScanExec is not redundant. When SparkPlan match 4 scenarios(1. [[ProjectExec]] -> [[FilterExec]] -> [[FileSourceScanExec]]...), the program need to match FileSourceScanExec first, then using OapFileSourceScanExec to replace it.

File: oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/CacheHotTablesSuite.scala (hunk 4 above)
Comment (haojinIntel) [reply oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/CacheHotTablesSuite.scala]: If not optimized the relation is FileSourceScanExec. So I create new functions to  judge the relation is FileSourceScanExec when not using OapFileSourceStrategy.

**Review by LuciferYang — APPROVED** [review]

**Pull request merged**